data_structures/linked_list.py

```
@@ -5,54 +5,61 @@
 
 class LinkedListIterator:
     """ Iterator for LinkedList. """
-    def __init__(self, head_node: Node[T], parent_list:LinkedList[T]):
-        self.__next = head_node
-        self.__curr = None
-        self.__prev = None
+    def __init__(self, parent_list:LinkedList[T]):
+        self.__curr:Node[T] | None = None
+        self.__prev:Node[T] | None = None
         self.__parent = parent_list
 
     def __iter__(self):
         return self
 
     def __next__(self) -> T:
-        if self.__next is None:
-            raise StopIteration
+        if self.__curr is None:
+            self.__curr = self.__parent._LinkedList__head
+            if self.__curr is None:
+                raise StopIteration
+            return self.__curr.item
+
         else:
-            item = self.__next.item
-            if self.__curr is not self.__next:
-                self.__prev = self.__curr
-            self.__curr = self.__next
-            self.__next = self.__next.link
-            return item
+            if self.__curr.link is None:
+                raise StopIteration
+            self.__prev = self.__curr
+            self.__curr = self.__curr.link
+            return self.__curr.item
         
     def insert_before(self, item:T):
         new_node = self.__parent._insert_after_node(item, self.__prev)
-        self.__prev = new_node
+        if self.__curr is None:
+            pass
+        else:
+            self.__prev = new_node
 
```

I'm not sure if inserting before 'before the first node of the List' makes sense, maybe line 32 should raise an error ( and check if `__curr` is none before adding)